Fix #160622: change containsWatchConpanion function to detect companion watch apps defined by only the project info file.

garrettjavalia opened the pull request:
(Please be generous for some typos and grammar error in sentences below. English is not my mother tongue)

TRDR : This patch fixes flutter run(and might be flutter build too) failure on iOS flutter apps with companion watchOS app in case the WKCompanionAppBundleIdentifier value is defined by xcode project configuration and uses different app bundle id by scheme(which happens when you do something like com.myapp.app1.dev things for scheme bundle id)

This change checks default scheme (the scheme with corresponding name for the debug/release mode and selected build flavor.) in the build settings check step of containsWatchCompanion function.


current code of containsWatchCompanion function works like below

1. check all default Info.plist file's content of all targets to determine if the project has watchOS companion apps.
(this doesn't work well in mordern xcode settings when they use multi plist files which are selected on build time by build configuration. Sometimes, the default Info.plist file doesn't even exist in the project.)

2. check if "WKCompanionAppBundleIdentifier" is included in the xcode project info file(the ios/Runner.xcodeproj/project.pbxproj file in case of iOS project generated by flutter)

3. If "WKCompanionAppBundleIdentifier" has found in the project info file,
check build configuration variables of every single scheme but the current target scheme if they have config value with key "INFOPLIST_KEY_WKCompanionAppBundleIdentifier" identical to current build configuration's bundle identifier returned by productBundleIdentifier@xcode_project.dart function.


I believe The third step causes many problems reported in some issue and pr, saying iOS app with companion watchOS app fails on build or run.
https://github.com/flutter/flutter/issues/160622
https://github.com/flutter/flutter/pull/172436

In my case, My iOS project had multiple schemes which are linked to different build configurations, but using single Runner setting per app.
because we exclude current scheme for the check, we get all the INFOPLIST_KEY_WKCompanionAppBundleIdentifier values except the one which should be same with the identifier returned by productBundleIdentifier@xcode_project.dart function.
that is because companion watch apps doesn't have to have same WKCompanionAppBundleIdentifier with other scheme's main app. We have to check the default scheme to see watchOS companion app`s WKCompanionAppBundleIdentifier, not other scheme. (other schemes can have slightly different or totally different bundle id. that was to support different environment systems like dev, prod in my case)
as a result, flutter run gave me WatchOS app built for device target iOS/iPad, not Watch. which caused problem during installation of iOS app to a simulator because the embeded Watch App was in invalid format. I found this by inspecting Info.plist of generated watchOS app.

below is part of the flutter run verbose log with failure caused by old code.
[+1154 ms] An error was encountered processing the command (domain=IXUserPresentableErrorDomain, code=1):
           App installation failed: ‘Bora Debug-dev’을(를) 설치할 수 없음 (which says "cannot install" in korean)
           나중에 다시 시도하십시오. (which says "try again later" in korean)
           Found WatchKit 2.0 app at /Users/javalia/Library/Developer/CoreSimulator/Devices/071691B3-7901-47E5-9B38-4D5B799F3530/data/Library/Caches/com.apple.mobile.installd.staging/temp.8HZIEG/extracted/Payload/Runner.app/Watch/bora Watch App.app but it does not have a WKWatchKitApp or WKApplication key set to true in its Info.plist
           Underlying error (domain=IXUserPresentableErrorDomain, code=1):
           	‘Bora Debug-dev’을(를) 설치할 수 없음 (which says "cannot install" in korean)
           	나중에 다시 시도하십시오. (which says "try it later" in korean)


I suggest this change with two reasons : 

1. I think forcing watchOS app makers split their runner configurations to support watch companion app is something not recommendable/not good to force as a convention.
(Which is implied in the comment which is removed by this pr, and could be one of the valid detouring.
some people in this issue https://github.com/flutter/flutter/issues/160622 said that a detouring is adding of dummy WKCompanionAppBundleIdentifier configuration in iOS app, which should be in only watchOS apps in normal.)

2. The function should work as it's name implies, so I guess it should not omit default scheme during check. Matching function's behavior with it's name will be likely to reduce potential errors caused by this change while fixing problems.

## Pre-launch Checklist

- [ X ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ X ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ X ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ X ] I signed the [CLA].
- [ X ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

garrettjavalia committed:
Remove skip for default scheme in scheme loop

Previously, the loop skipped the default scheme when iterating over project schemes. This commit removes that check, ensuring all schemes, including the default, are processed.

garrettjavalia committed:
Change containsWatchCompanion to check only active scheme

garrettjavalia committed:
Check only default scheme for watch companion bundle ID

Refactors the logic to evaluate only the default scheme's build settings when verifying the watch companion app bundle identifier, instead of iterating through all schemes.

garrettjavalia committed:
Merge branch 'master' into containsWatchCompanion-fix

garrettjavalia committed:
Merge branch 'master' into containsWatchCompanion-fix

vashworth commented:
It appears that this PR checks for `WKCompanionAppBundleIdentifier` within the build settings of the default scheme. However, this does not makes sense as `WKCompanionAppBundleIdentifier` should be set in the companion Watch app's settings, not the iOS app target settings: https://developer.apple.com/documentation/bundleresources/information-property-list/wkcompanionappbundleidentifier?language=objc#Discussion

@garrettjavalia Did this solution not work for you? https://github.com/flutter/flutter/pull/172436

garrettjavalia commented:
I did not test solution in #172436. My solution was made separately from that, and I see that solution is currently closed.

please let me explain about this more.

first of all, checking default(currently targeting) scheme is a valid operation in my opinion.
schemes can include multiple build targets(the apps) and build configurations to use on each operation(debug run, build, etc.)
Imagine a valid Xcode scheme which includes two targets(a host iOS app and a companion watchOS app.) for archive.(which flutter build command should handle, in real world senarios.)
(we surely can omit watchOS app in the scheme we want to use in flutter run command, but not for the flutter build command.)

the second part, which you might feel this code is wrong is, How it can work if it just processes xcode build configuration without specifying correct watchOS app as target. (buildSettingsForBuildInfo function and it`s call chain can specify a target to test, but current code doesn't specify one.)
by running flutter run -v, we can see that if no target is given, it generates config for all targets in the project at once.
and it surely includes INFOPLIST_KEY_WKCompanionAppBundleIdentifier if the scheme has watchOS app which is correctly configured by xcode project's target build settings.

now we can think that the valid way to handle this problem might be by making it check only watchOS App's build configuration, and yes that is a valid point. but that is for some cases when a scheme has two watchOS app(like one for companion app and one for independent app, or just two companion watchOS apps in wrong configuration.)

with above inspections, I suggest current PR a pretty good way to solve current situation. well, maybe I can improve the code to handle more complicated situations I mentioned above, but it would be optional for now

garrettjavalia commented:
Following screenshot shows how a scheme can have multiple apps in it's scheme configuration of xcode.
<img width="954" height="516" alt="스크린샷 2025-10-25 오후 6 45 32" src="https://github.com/user-attachments/assets/fe8db423-b97f-4a83-8cb2-bcbb869c24be" />

And this shows how a host iOS app handles it's companion watchOS app during the build phases.
<img width="1100" height="935" alt="스크린샷 2025-10-25 오후 6 47 15" src="https://github.com/user-attachments/assets/d4471931-db57-4f8b-8024-fb4f989434b5" />

So in short, Xcode project scheme != Xcode build target(app)

garrettjavalia committed:
Check all Xcode schemes for watch companion app bundle ID

Previously, only the default scheme was checked for the WKCompanionAppBundleIdentifier. This update iterates through all available schemes in the project to ensure the correct bundle identifier is found, improving robustness against scheme with multiple watchOS apps.

garrettjavalia committed:
Fix containsWatchCompanion function test cases and add test cases with only default scheme

garrettjavalia commented:
Updated existing tests and added new tests as a separated test group.

garrettjavalia committed:
Refactor test setup for 'watch companion' test group

Moved the initialization of mockXcodeProjectInterpreter's buildSettingsByBuildContext for the 'Runner' scheme from individual test cases and group setUp blocks to a shared setUp block.

garrettjavalia committed:
Fix iOS project test setup for Xcode variable substitution

Adds mock build settings for XcodeProjectBuildContext in two test cases to ensure correct variable substitution during plist parsing.

garrettjavalia commented:
found out that 'watch companion' test group's test codes are getting host iOS app bundle name configuration with XcodeProjectBuildContext(scheme: 'Runner')
while it requires XcodeProjectBuildContext(scheme: 'Runner',  deviceId: '123' ) for substituteXcodeVariables calls used to parse PList file.

refactored test codes in following ways.
1. Removed ineffective configurations related to substituteXcodeVariables in test cases those are not interacting with plist xcode variable.
2. unified code that sets bundle id into the 'watch companion' test group, which were divided into sub testgroups in my previous commit.

garrettjavalia commented:
I found that multi target schemes are Ok in xcode, but not in flutter code.
productBundleIdentifier function in xcode_project.dart is weak to multiple targets so that it can detect wrong bundle identifier.
It was pure luck that my xcode project was working fine in flutter run and flutter build command.
I will think of improvement to this. until then, I guess this PR should be draft

garrettjavalia committed:
remove support for watch companion app detection with single scheme project

Detection of watch companion app without dedicated watch app target scheme requires watch companion app target included as a explicit build target of the default scheme, which causes malfunctioning of bundle identifier detection logic of flutter ios build chain which assumes default scheme will have only one explicit build target which generates it's build configuration for xcodebuild command.

hellohuanlin committed:
Merge branch 'master' into containsWatchCompanion-fix

LouiseHsu commented on packages/flutter_tools/lib/src/xcode_project.dart:
```diff
@@ -796,13 +796,13 @@ def __lldb_init_module(debugger: lldb.SBDebugger, _):
     }
     for (final String scheme in projectInfo.schemes) {
-      // the default scheme should not be a watch scheme, so skip it
+      // Flutter assumes single build target per scheme, so skip defalt scheme.
```
nit: defalt -> default

LouiseHsu reviewed (approved):
tiny typo nit, lgtm!

garrettjavalia committed:
fix a typo in comment.

garrettjavalia committed:
Merge branch 'master' into containsWatchCompanion-fix
